### PR TITLE
Add the ability to reset & change the password of the personal account

### PR DESCRIPTION
**Pull request by Anastasia-Troian**

### CODE REVIEW

File: Backend/Degree project/Controllers/AuthenticationController.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using Core.DTO.User;
 using Core.Helpers;
 using Core.Interfaces.CustomService;
+using Core.Services;
```
Comment (marynakernychna) [inline Backend/Degree project/Controllers/AuthenticationController.cs]: Needless

File: Backend/Degree project/Controllers/AuthenticationController.cs [hunk 2]
```diff
@@ -48,6 +52,17 @@ public async Task<IActionResult> LogoutAsync([FromBody] UserLogoutDTO userLogout
             return Ok();
         }
 
+        [HttpPut("change-password")]
+        [AuthorizeByRole(IdentityRoleNames.User)]
+        public async Task<IActionResult> ChangePasswordAsync(ChangePasswordDTO changePasswordDTO)
```
Comment (marynakernychna) [inline Backend/Degree project/Controllers/AuthenticationController.cs]: [FromBody]

File: Backend/Core/Services/AuthenticationService.cs [hunk 3]
```diff
@@ -10,6 +10,7 @@
 using Core.Resources;
 using Core.Specifications;
 using Microsoft.AspNetCore.Identity;
+using RTools_NTS.Util;
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: ???

File: Backend/Core/Services/AuthenticationService.cs [hunk 4]
```diff
@@ -161,6 +162,17 @@ public async Task ResetPasswordAsync(ResetPasswordDTO resetPasswordDTO)
                         ErrorMessages.NewInfoSamePrevious,
                         HttpStatusCode.BadRequest);
             }
+            if (!await this._userManager.VerifyUserTokenAsync(
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: Add a blank line

Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: this?

File: Backend/Core/Services/AuthenticationService.cs [hunk 5]
```diff
@@ -161,6 +162,17 @@ public async Task ResetPasswordAsync(ResetPasswordDTO resetPasswordDTO)
...
+                user, 
+                _userManager.Options.Tokens.PasswordResetTokenProvider, 
+                "ResetPassword", 
+                resetPasswordDTO.Token))
+            {
+
+                throw new HttpException(
+                        "Token Invalid",
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: We have error messages

File: Backend/Core/Services/AuthenticationService.cs [hunk 6]
```diff
@@ -161,6 +162,17 @@ public async Task ResetPasswordAsync(ResetPasswordDTO resetPasswordDTO)
...
+                        HttpStatusCode.BadRequest);
+            }
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: We did u add this? The ResetPasswordAsync method'll throw this error

File: Frontend/src/components/modals/changePassword/index.jsx [hunk 7]
```diff
@@ -52,6 +52,7 @@ function ChangePasswordModal(props) {
             cancelButtonProps={{ style: { display: 'none' } }}
             okButtonProps={{ style: { display: 'none' } }}
             footer={null}
+            width={750}
```
Comment (marynakernychna) [inline Frontend/src/components/modals/changePassword/index.jsx:47]: U can do this in CSS, can't u?

File: Backend/Core/Services/AuthenticationService.cs [hunk 8]
```diff
@@ -109,6 +110,39 @@ public async Task LogoutAsync(UserLogoutDTO userLogoutDTO)
             await _refreshTokenRepository.DeleteAsync(refreshToken);
         }
 
+        public async Task ChangePasswordAsync(ChangePasswordDTO changePasswordDTO, string userId)
+        {
+            if (changePasswordDTO.CurrentPassword == changePasswordDTO.NewPassword)
+            {
+                throw new HttpException(
+                        ErrorMessages.NewInfoSamePrevious,
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs:117]: Use another message

**Review by marynakernychna — COMMENTED** [review]
This message doesn't fit 'cause I didn't change anything, but entered everything from scratch

image

"... password..."

image

File: Backend/Core/Services/AuthenticationService.cs (hunk 6 above)
Comment (Anastasia-Troian) [reply Backend/Core/Services/AuthenticationService.cs]: The ResetPasswordAsync method does not throw an error if the token has not been used

File: Frontend/src/components/modals/changePassword/index.jsx (hunk 7 above)
Comment (Anastasia-Troian) [reply Frontend/src/components/modals/changePassword/index.jsx:47]: via SCSS does not work

File: Backend/Core/Resources/ErrorMessages.resx [hunk 9]
```diff
@@ -153,6 +153,9 @@
   <data name="NewInfoSamePrevious" xml:space="preserve">
     <value>The new information is the same as the previous one!</value>
   </data>
+  <data name="NewPasswordSameCurrentPassword" xml:space="preserve">
+    <value>New password same as current password!</value>
```
Comment (marynakernychna) [inline Backend/Core/Resources/ErrorMessages.resx]: Where's the article?

File: Backend/Core/Resources/ErrorMessages.Designer.cs [hunk 10]
```diff
@@ -168,6 +168,15 @@ internal class ErrorMessages {
             }
         }
         
+        /// <summary>
+        ///   Looks up a localized string similar to New password same as current password!.
+        /// </summary>
+        internal static string NewPasswordSameCurrentPassword {
+            get {
+                return ResourceManager.GetString("NewPasswordSameCurrentPassword", resourceCulture);
```
Comment (marynakernychna) [inline Backend/Core/Resources/ErrorMessages.Designer.cs]: AS

File: Backend/Core/Services/AuthenticationService.cs [hunk 11]
```diff
@@ -115,7 +114,7 @@ public async Task ChangePasswordAsync(ChangePasswordDTO changePasswordDTO, strin
             if (changePasswordDTO.CurrentPassword == changePasswordDTO.NewPassword)
             {
                 throw new HttpException(
-                        ErrorMessages.NewInfoSamePrevious,
+                        ErrorMessages.NewPasswordSameCurrentPassword,
```
Comment (marynakernychna) [inline Backend/Core/Services/AuthenticationService.cs]: I don't like the two options, but if I had to choose between the two, the first one is better

**Review by marynakernychna — APPROVED** [review]

**Pull request merged**